accord-core/src/test/java/accord/impl/basic/Cluster.java

```diff
@@ -219,18 +220,50 @@ public void processAll()
             processNext(next);
     }
 
+    boolean hasNonRecurring()
+    {
+        boolean hasNonRecurring = false;
+        for (Pending p : pending)
+        {
+            if (!(p instanceof RecurringPendingRunnable))
+                continue;
+            RecurringPendingRunnable r = (RecurringPendingRunnable) p;
+            if (r.requeue.hasNonRecurring())
+            {
+                hasNonRecurring = true;
+                break;
+            }
+        }
+
+        return hasNonRecurring;
+    }
+
     public boolean processPending()
     {
         checkFailures.run();
-        if (pending.size() == recurring)
+        // All remaining tasks are recurring
+        if (recurring > 0 && pending.size() == recurring && !hasNonRecurring())
             return false;
 
-        Object next = pending.poll();
+        Pending next = pending.poll();
         if (next == null)
             return false;
 
         processNext(next);
+
         checkFailures.run();
+
+        // If drain was requested, run until we have no non-recurring tasks are available
+        if (next instanceof RecurringPendingRunnable && ((RecurringPendingRunnable) next).drainAfterRunning)
```

Comment:
Might it be cleaner to submit a kind of "drain" task?

Reply:
by which I mean, instead of making the recurring task say "drain after me", the task submits a one-shot drain task whenever it wants to perform a drain?

Reply:
This drain task could itself submit some other post-drain task to e.g. re-enable messaging (and schedule the next journal replay task)